pulsar-broker-auth-oidc/src/main/java/org/apache/pulsar/broker/authentication/oidc/OpenIDProviderMetadataCache.java

```diff
@@ -0,0 +1,124 @@
+                            future.completeExceptionally(new AuthenticationException(
+                                    "Error retrieving OpenID Provider Metadata at " + issuer + ": " + e.getMessage()));
+                        }
+                        return future;
+                    });
+    }
+
+    private final AsyncLoadingCache<String, OpenIDProviderMetadata> cache;
+
+    OpenIDProviderMetadataCache(ServiceConfiguration config, AsyncHttpClient httpClient) {
+        int maxSize = getConfigValueAsInt(config, CACHE_SIZE, CACHE_SIZE_DEFAULT);
+        int expireAfterSeconds = getConfigValueAsInt(config, CACHE_EXPIRATION_SECONDS,
+                CACHE_EXPIRATION_SECONDS_DEFAULT);
+        this.cache = Caffeine.newBuilder()
+                .maximumSize(maxSize)
+                .expireAfterWrite(expireAfterSeconds, TimeUnit.SECONDS)
+                .buildAsync(getLoader(httpClient));
+    }
+
+    /**
+     * Retrieve the OpenID Provider Metadata for the provided issuer.
+     * <p>
+     * Note: this method does not do any validation on the parameterized issuer. The OpenID Connect discovery
+     * spec requires that the issuer use the HTTPS scheme: https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata.
+     * The {@link AuthenticationProviderOpenID} class handles this verification.
+     *
+     * @param issuer - authority from which to retrieve the OpenID Provider Metadata
+     * @return the {@link OpenIDProviderMetadata} for the given issuer
+     * @throws AuthenticationException if any exceptions occur while retrieving the metadata.
+     */
+    CompletableFuture<OpenIDProviderMetadata> getOpenIDProviderMetadataForIssuer(@Nonnull String issuer) {
+        return cache.get(issuer);
+    }
+
+    /**
+     * Verify the issuer url, as required by the OpenID Connect spec:
+     *
+     * Per the OpenID Connect Discovery spec, the issuer value returned MUST be identical to the
+     * Issuer URL that was directly used to retrieve the configuration information. This MUST also
+     * be identical to the iss Claim value in ID Tokens issued from this Issuer.
+     * https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderConfigurationValidation
+     *
+     * @param issuer - the issuer used to retrieve the metadata
+     * @param metadata - the OpenID Provider Metadata
+     * @throws AuthenticationException if the issuer does not exactly match the metadata issuer
+     */
+    private void verifyIssuer(@Nonnull String issuer, OpenIDProviderMetadata metadata) throws AuthenticationException {
+        if (!issuer.equals(metadata.getIssuer())) {
+            incrementFailureMetric(AuthenticationExceptionCode.ISSUER_MISMATCH);
+            throw new AuthenticationException(String.format("Issuer URL mismatch: [%s] should match [%s]",
+                    issuer, metadata.getIssuer()));
+        }
+    }
```

Comment:
I would suggest that this validation not be performed, because the Kubernetes OpenID Discovery feature is not 100% spec-compliant.  In particular, the actual `issuer` value may be different from the configured value.  For example, here's the disco document from a GKE cluster:

```
$ curl -k https://kubernetes.default.svc/.well-known/openid-configuration
{
    "issuer": "https://container.googleapis.com/v1/projects/xyz/locations/us-central1/clusters/test-oidc",
    "jwks_uri": "https://35.223.170.12:443/openid/v1/jwks",
    "response_types_supported": [
        "id_token"
    ],
    "subject_types_supported": [
        "public"
    ],
    "id_token_signing_alg_values_supported": [
        "RS256"
    ]
}
```

Reply:
That's an interesting point of divergence. I noticed somewhat similar issues with keycloak where the `iss` claim on the token would change depending the DNS name you used to retrieve the token. In that case, I just added multiple issuers to ensure all possible names were trusted.

> Note that, when an ID token arrives at the broker, the `iss` claim will point to the actual value (`https://container.googleapis.com/v1/projects/xyz/....`).

If the ID token's `iss` claim matches the `issuer` on the discovery document, do we need to worry that Kubernetes provides another way to get the discovery document?

Reply:
I take your comment to be about whether to allow for the broker configuration to simply use the well-known value of `https://kubernetes.default.svc/.well-known/openid-configuration` as an issuer URL, as opposed to the dynamic value as seen in the metadata doc. I would advocate for supporting the well-known value for simplicity's sake.  I'd love to have a reusable broker config with in-built Kubernetes authentication.

Reply:
I agree that using `https://kubernetes.default.svc/.well-known/openid-configuration` enhances simplicity. I decided to prototype it using the kubernetes `ApiClient` since that takes it a step further and discovers the URL, the relevant root CA, and the appropriate authentication.